---
title: "Gopalakrishnan Demux Dada"
author: "Katherine Mueller"
date: "7/30/2021"
output: html_document
---

#Load Libraries
```{r}
library(readr)
library(fs)
library(dplyr)
library(tibble)
library(Biostrings)
library(dada2)
library(stringr)
library(magrittr)
library(ggplot2)
library(tidyr)
library(phyloseq)
```

#Set up Paths, Directories, and Shell Variables
```{r}
#Directories
data.dir = "/work/kdm65/gopalakrishnan"
output.dir = "/work/kdm65/scratch"
```

```{r}
if (dir_exists(output.dir)) {
  dir_delete(output.dir)
}
dir_create(output.dir)
```

```{r}
demux.dir = file.path(output.dir, "demux")

#Files
map.file = file.path(data.dir, "../Gopalakrishnan_map.txt")
silva.ref = file.path(data.dir, "../silva_nr99_v138.1_wSpecies_train_set.fa.gz")
ps.rds = file.path(data.dir, "../gopalakrishnan.rds")
sample.ids = file.path(data.dir, "sample_id.txt")

#Bash variables
Sys.setenv(RAW_FASTQ_DIR = data.dir)
```

#Demultiplexing
```{bash}
run_path=/work/kdm65/gopalakrishnan
cd $run_path
for sam in `less sample_id.txt`
do
    echo $sam
    less /work/kdm65/gopalakrishnan/ERR2162225.fastq | grep $sam -A 3 > $sam.fastq
done
```

```{r}
if (dir_exists(demux.dir)) {
  dir_delete(demux.dir)
}
dir_create(demux.dir)
```

```{bash}
cd $RAW_FASTQ_DIR
ls Wargo*
cp Wargo* /work/kdm65/scratch/demux
```

#Filter and Trim
Get lists of reads
```{r}
fnFs <- sort(list.files(demux.dir, pattern = ".fastq", full.names = TRUE))

forward_fastq_suffix = ".fastq"

fnFs %>%
  basename %>%
  str_replace(forward_fastq_suffix,"") ->
  sample.names
```

```{r}
print(fnFs)
```

```{r}
print(sample.names)
```

#Quality Profiles
```{r}
plotQualityProfile(fnFs[1:2])
```

The paper said that these were 2x250 paired end reads. I take it the 250-500 are related to the reverse reads, but I don't know how to handle this besides just using the forward portions.

#Filter Reads
Assign filepaths for filtered files
```{r}
filt_path <- file.path(output.dir, "filtered")
filtFs <- file.path(filt_path, paste0(sample.names, "_filt.fastq.gz"))
```

Filter Reads
```{r}
filt.out <- filterAndTrim(fnFs, filtFs, trimLeft = 10, truncLen = 240,
                           maxN = 0, maxEE = 2, truncQ = 2, rm.phix = TRUE,
                           compress = TRUE, multithread = FALSE)
```

```{r}
head(filt.out)
```

#Learn Error Rates
```{r}
errF <- learnErrors(filtFs, multithread = FALSE)
```

```{r}
plotErrors(errF, nominalQ = TRUE)
```

#Dereplication
```{r}
derepFs <- derepFastq(filtFs, verbose = TRUE)
names(derepFs) <- sample.names
```

#Sample Inference
```{r}
dadaFs <- dada(derepFs, err=errF, multithread = FALSE)
```

```{r}
dadaFs[[2]]
```

#Further Processing
Construct sequence table
```{r}
seqtab <- makeSequenceTable(dadaFs)
dim(seqtab)
```

```{r}
table(nchar(getSequences(seqtab)))
```

Remove chimeras
```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab, method = "consensus", multithread = FALSE, verbose = TRUE)
dim(seqtab.nochim)
```

```{r}
sum(seqtab.nochim)/sum(seqtab)
```

Track reads through the pipeline
```{r}
getN <- function(x) sum(getUniques(x))
filt.out %>%
  as_tibble(rownames = "filename") %>%
  mutate(sample = str_replace(filename, forward_fastq_suffix,"")) %>%
  select(sample, input = reads.in, filtered = reads.out) ->
  track

sapply(dadaFs, getN) %>%
  enframe(name = "sample", value = "denoised") ->
  denoised
track %<>% full_join(denoised, by = c("sample"))

rowSums(seqtab) %>%
  enframe(name = "sample", value = "tabled") ->
  tabled
track %<>% full_join(tabled, by = c("sample"))

rowSums(seqtab.nochim) %>%
  enframe(name = "sample", value = "nonchim") ->
  nonchim
track %<>% full_join(nonchim, by = c("sample"))

track
```

```{r}
track %>%
  gather(key = "stage", value = "counts", -c("sample")) %>%
  replace_na(list(counts = 0)) %>%
  mutate(stage=factor(stage, levels = c('input', 'filtered', 'denoised', 'tabled', 'nonchim'))) %>%
  ggplot(mapping = aes(x = stage, y = counts, by = sample, group = sample)) + geom_line(alpha = 0.05) + theme_classic()
```

#Assign Taxonomy
```{r}
taxa <- assignTaxonomy(seqtab.nochim, silva.ref, multithread = FALSE)
taxa.print <- taxa
rownames(taxa.print) <- NULL
head(taxa.print)
```

#Make Phyloseq Object
Load metadata
```{r}
metadata.df = read_tsv(map.file) %>%
  dplyr::rename(Sample = "#SampleID") %>%
  column_to_rownames("Sample") %>%
  as.data.frame()

metadata.df
```

Construct phyloseq object
```{r}
otus = otu_table(seqtab.nochim, taxa_are_rows = FALSE)
sd = sample_data(metadata.df)
ps <- phyloseq(otus, sd, tax_table(taxa))

ps
```

Save phyloseq object as RDS
```{r}
write_rds(ps, ps.rds)
```

Confirm that the RDS is usable
```{r}
loaded.ps = read_rds(ps.rds)
print(loaded.ps)
```

#Session Info
```{r}
sessionInfo()
```

